### DISCUSSION TITLE
OPC UA integration into EDC

### DISCUSSION THREAD

**Discussion by ma3u** [eclipse-edc/Connector#2875] [Ideas]
We discussed how manufacturer can use the EDC for the existing communication technology [OPC-UA](https://opcfoundation.org/about/opc-technologies/opc-ua/) on the shop floor.

Catena-X is using EDC as central component to connect the factories and backend systems in the complex and global supply chain with each other. In the [documentation ](https://catena-x.net/en/offers/edc-the-central-component) you will find further information.  for interoperability. 

OPC UA is supporting various protocols like HTTP, TCP, MQTT as this [article ](https://opcconnect.opcfoundation.org/2021/12/opc-ua-mqtt-and-information-interoperability/) describes. OPC UA is based on the [Subscribe and Publish communication](https://opcconnect.opcfoundation.org/2021/12/opc-ua-publisher-subscriber-in-cloud-based-solutions/) pattern. The EDC data plane OPC UA extension need to subscribe for one or more machines for specific data services. The data will be pushed by the OPC data provider to the OPC data consumer.  

The operators want to connect their machines of the shop floor with the service provides for predictive maintenance, trace of carbon footprint, quality assurance, etc. 

Data offerings need to be grouped over machines. The first idea was to establish multi-tendency: https://github.com/eclipse-edc/Connector/discussions/1301 @jimmarino make clear:  “A fundamental design principle of EDC is that a participant agent is associated with one identity.” … “Run each EDC as a ReplicaSet and configure Kubernetes to perform routing, isolation, and other requirements. The ReplicaSet run on a shared compute infrastructure and map URLs to different Kubernetes ingress points.”

So each machine has an own identity that means also an own EDC instance. The EDC instance also provide a link to a [federated catalog based on W3C DCAT](https://docs.internationaldataspaces.org/dataspace-protocol/catalog/catalog.protocol). The catalog provides a [dcat:dataset](https://www.w3.org/TR/vocab-dcat-3/#Class:Dataset) object. The category of provided data be set in the [dct:theme](https://www.w3.org/TR/vocab-dcat-3/#Property:resource_theme). 

If the shop floor already provide a service that aggregates this data from various machines for a common use case like PCF calculation, this data can be provided as own data service in a separate EDC instance.Usually a HTTP Rest API provides the calculated results. 

The OPC UA will be used in data spaces like EuroProGigant, Catena-X and Manufacturing-X.

<img width="100%" alt="image" src="https://user-images.githubusercontent.com/17936678/235891108-51dd7496-f323-4bf0-b6a9-ed295c32e514.png">

**Implementation hint:**

The Digital Twin Consortium already implemented a **OPC UA PubSub Service** [OPCUAPubSubService.cs](https://github.com/digitaltwinconsortium/AAS-Repository/blob/masterV3/src/Services/OPCUAPubSubService.cs) in C#. This analyze the PubSub JSON log output via Cloud Analytics Service [Azure Data Explorer](https://learn.microsoft.com/en-us/azure/data-explorer/kusto/query/) (ADX).  Microsoft also provide [ADX SDK for Java](https://github.com/Azure/azure-kusto-java). 

If you don't want to have the dependency to a specific cloud service like ADX the EDC data plane can subscribe on the MQTT Server for a specific topic. The MQTT Server will push the updates as JSON to the provided endpoint of the EDC. There is also an implementation [UAPubSubMessageProcessor.cs](https://github.com/OPCFoundation/UA-CloudDashboard/blob/main/UAPubSubMessageProcessor.cs) you can reuse.  The OPC UA provides a Java client SDK called [UA Java](http://opcfoundation.github.io/UA-Java-Legacy/) to communicate with the OPC UA publisher and parse the messages.  

EDC provide a push mechanism to get proactively data via HTTP. See also [Provider Push documentation](https://github.com/eclipse-edc/Connector/tree/main/extensions/control-plane/transfer/transfer-data-plane#provider-push-1)

<img width="100%" src="https://raw.githubusercontent.com/eclipse-edc/Connector/main/docs/developer/architecture/data-transfer/diagrams/data-flow-http-push.png">

You also find an [HTTP Push example in the GitHub repository](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-07-provider-push-http).

**Comment by ma3u** [comment]
OPC Foundation is working on an integration together with T-Systems. So we will continue the discussion.

**Reply by GoetzGoerisch** [reply]
Looking forward how this works out.

**Reply by nicir** [reply]
are there any news?
